src/node_encoding.cc

```diff
@@ -0,0 +1,136 @@
+#include "util-inl.h"
+#include "v8-fast-api-calls.h"
+#include "v8.h"
+
+namespace node {
+
+using v8::ArrayBuffer;
+using v8::BackingStore;
+using v8::CFunction;
+using v8::Context;
+using v8::FastApiTypedArray;
+using v8::FastApiCallbackOptions;
+using v8::FastOneByteString;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::String;
+using v8::Uint8Array;
+using v8::Value;
+
+namespace encoding_methods {
+
+void GetUtf8Length(const v8::FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  CHECK_GE(args.Length(), 1);
+  CHECK(args[0]->IsString());
+
+  args.GetReturnValue().Set(args[0].As<String>()->Utf8Length(isolate));
+}
+
+int FastGetUtf8Length_(const char* str) {
+  int length = 0;
+  for (; *str; ++str) if (*str & 0x80) length += 2; else length += 1;
+  return length;
+}
+
+int32_t FastGetUtf8Length(Local<Value> receiver, const FastOneByteString& source) {
+  return FastGetUtf8Length_(source.data);
+}
+
+void Encode(Isolate* isolate, Local<String> str, Local<Uint8Array> out) {
+  Local<ArrayBuffer> buf = out->Buffer();
+  size_t out_length = out->Length();
+  char* write_result = static_cast<char*>(buf->Data()) + out->ByteOffset();
+
+  str->WriteUtf8(isolate,
+                 write_result,
+                 out_length,
+                 nullptr,
+                 String::NO_NULL_TERMINATION | String::REPLACE_INVALID_UTF8);
+}
+
+void EncodeUtf8(const v8::FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  CHECK_GE(args.Length(), 2);
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsUint8Array());
+
+  Local<String> str = args[0].As<String>();
+  Local<Uint8Array> out = args[1].As<Uint8Array>();
+  size_t utf8_length = str->Utf8Length(isolate);
+
+  if (utf8_length != out->Length()) {
+    auto buf = ArrayBuffer::New(isolate, utf8_length);
+    auto correct_out = Uint8Array::New(buf, 0, utf8_length);
+    Encode(isolate, str, correct_out);
+    args.GetReturnValue().Set(correct_out);
+    return;
+  }
+
+  Encode(isolate, str, out);
+  args.GetReturnValue().Set(out);
+}
+
+void latin1_to_utf8(const char* str, char* c) {
+  for (; *str; ++str) {
+    if (*str & 0x80) {
+      *c++ = *str;
+    } else {
+      *c++ = (char)(0xc0 | (unsigned)*str >> 6);
+      *c++ = (char)(0x80 | (*str & 0x3f));
+    }
+  }
+  *c++ = '\0';
+}
+
+void FastEncodeUtf8(Local<Value> receiver,
+                    const FastOneByteString& source,
+                    const FastApiTypedArray<uint8_t>& output,
+                    FastApiCallbackOptions& options) {
+  size_t utf8_length = FastGetUtf8Length_(source.data);
+  uint8_t* storage;
+  auto is_available = output.getStorageIfAligned(&storage);
+
+  if (is_available && utf8_length == output.length()) {
+    latin1_to_utf8(source.data, reinterpret_cast<char*>(storage));
+  } else {
+    options.fallback = true;
+  }
```

Comment:
you can set `options.fallback` at any point in execution, which means you can opportunistically start copying the bytes over and only bail out when you encounter multi-byte sequences, rather than checking the length up front.

Reply:
That makes sense. Thanks! Do you know any existing implementation for latin1 to utf8 in node or its dependencies?